src/Symfony/Contracts/Service/ServiceSubscriberTrait.php

```diff
@@ -36,13 +37,65 @@ public static function getSubscribedServices(): array
+            foreach ((new \ReflectionClass(self::class))->getMethods() as $method) {
+                if (self::class !== $method->getDeclaringClass()->name) {
+                    continue;
+                }
+
+                if (!$attribute = $method->getAttributes(SubscribedService::class)[0] ?? null) {
+                    continue;
+                }
+
+                if ($method->isStatic() || $method->isAbstract() || $method->isGenerator() || $method->isInternal() || $method->getNumberOfRequiredParameters()) {
+                    // todo, what exception to throw?
+                    throw new \LogicException(sprintf('Cannot use "%s" on method "%s::%s()" (can only be used on non-static, non-abstract methods with no parameters).', SubscribedService::class, self::class, $method->name));
+                }
+
+                if (!$returnType = $method->getReturnType()) {
+                    // todo, what exception to throw?
+                    throw new \LogicException(sprintf('Cannot use "%s" on methods without a return type in "%s::%s()".', SubscribedService::class, $method->name, self::class));
+                }
+
+                if (!$returnType instanceof \ReflectionNamedType) {
+                    // todo, what exception to throw?
+                    throw new \LogicException(sprintf('Cannot use "%s" on methods with a union return type in "%s::%s()".', SubscribedService::class, $method->name, self::class));
+                }
+
+                if ($returnType->isBuiltin()) {
+                    // todo, what exception to throw?
+                    throw new \LogicException(sprintf('Cannot use "%s" on methods with a native return type in "%s::%s()".', SubscribedService::class, $method->name, self::class));
+                }
+
+                $services[$attribute->newInstance()->key ?? self::class.'::'.$method->name] = '?'.$returnType->getName();
+                $attributeOptIn = true;
             }
+        }
+
+        if (!$attributeOptIn) {
+            foreach ((new \ReflectionClass(self::class))->getMethods() as $method) {
+                if ($method->isStatic() || $method->isAbstract() || $method->isGenerator() || $method->isInternal() || $method->getNumberOfRequiredParameters()) {
+                    continue;
+                }
+
+                if (self::class !== $method->getDeclaringClass()->name) {
+                    continue;
+                }
+
+                if (!$returnType = $method->getReturnType()) {
+                    continue;
+                }
+
+                if ($returnType->isBuiltin()) {
+                    continue;
+                }
+
+                if (\PHP_VERSION_ID >= 80000) {
+                    trigger_deprecation('symfony/service-contracts', '5.4', 'Using the "%s" trait in "%s" without subscribed service methods marked with the "%s" attribute is deprecated and will not be allowed in Symfony 6.0.', ServiceSubscriberTrait::class, self::class, SubscribedService::class);
```

Comment:
Is `5.4` correct here, as the contracts does not follow the Symfony version numbers.
Current stable one is `2.4.0`

Reply:
Good point, what exactly triggers a new major contracts version? Will 3.0 be released with Symfony 6?

Reply:
Yes, contracts 3.0 will be released with Symfony 6.0